NKL/Encode.lean

```diff
@@ -0,0 +1,331 @@
+
+private def tag (t : UInt8) : List ByteArray -> ByteArray :=
+  List.foldl ByteArray.append (.mk #[t])
+
+private def encInt (i : Int) : ByteArray :=
+  match i with
+  | .ofNat n => tag 0 [encNat n]
+  | .negSucc n => tag 1 [encNat n]
+
+private def decInt : DecodeM Int := do
+  let sign <- next
+  let n <- decNat
+  bif sign == 0 then
+    return (.ofNat n)
+  else
+    return (.negSucc n)
+
+private def decInt' : DecodeM (Bool × Nat) := do
+  let sign <- next
+  let n <- decNat
+  return (sign == 1, n + sign.toNat)
+
+local instance : BEq ByteArray where
+  beq a b := a.data == b.data
+
+#guard encInt 1    == .mk #[0, 1, 0, 0, 0, 0, 0, 0, 0]
+#guard encInt 0    == .mk #[0, 0, 0, 0, 0, 0, 0, 0, 0]
+#guard encInt (-1) == .mk #[1, 0, 0, 0, 0, 0, 0, 0, 0]
+#guard encInt (-2) == .mk #[1, 1, 0, 0, 0, 0, 0, 0, 0]
+
+#guard decode' decInt (encInt 1) == some 1
+#guard decode' decInt (encInt 0) == some 0
+#guard decode' decInt (encInt $ -1) == some (-1)
+
+#guard decode' decInt' (encInt 1) == some (false, 1)
+#guard decode' decInt' (encInt 0) == some (false, 0)
+#guard decode' decInt' (encInt (-0)) == some (false, 0)
+#guard decode' decInt' (encInt $ -1) == some (true, 1)
+
+------------------------------------------------------------------------------
+-- Float values are encoded as (mantissa, exponent), where the
+-- mantissa and exponenet are integers.
+-- TODO:
+-- I am keeping the encode functions pure; encFloat can fail,
+-- but this seems unlikely as the Lean parser would have to not understand
+-- the result of toString on a Float value
+
+private def encFloat (f : Float) : ByteArray :=
+  let str := (toString f).toSubstring
+  let one : String.Pos := { byteIdx := 1 }
+  let sgn := str.get 0 == '-'
+  let str := bif sgn then { str with startPos := one } else str
+  match Lean.Syntax.decodeScientificLitVal? str.toString with
+  | none => panic! s!"Failed to parse {toString f}"
+  | some (m, esn, e) =>
+    let m := toInt sgn m
+    let e := toInt esn e
+    (encInt m).append (encInt e)
+where
+  -- this is to handle negative zero
+  toInt : Bool -> Nat -> Int := fun
+   | false, n => .ofNat n
+   | true,  0 => .ofNat 0
+   | true,  n => .negSucc (n-1)
+
+private def decFloat : DecodeM Float := do
+  let (sgn,m) <- decInt'
+  let (esn,e) <- decInt'
+  let f := Float.ofScientific m esn e
+  return bif sgn then -f else f
+
+#guard decode' decFloat (encFloat 0) == some 0
+#guard decode' decFloat (encFloat 0.5) == some 0.5
+#guard decode' decFloat (encFloat 1.0) == some 1.0
+#guard decode' decFloat (encFloat 1.5) == some 1.5
+#guard decode' decFloat (encFloat (-0)) == some (0.0)
+#guard decode' decFloat (encFloat (-0.5)) == some (-0.5)
+#guard decode' decFloat (encFloat (-1.0)) == some (-1.0)
+#guard decode' decFloat (encFloat (-1.5)) == some (-1.5)
+
+------------------------------------------------------------------------------
+-- Strings are encoded as a length followed by a sequence of UTF8 bytes
```

Comment:
Just reading the comment, I wonder how do you know how many bytes the length is?  I think this is bc you are using Nat in place of UInt64?

Reply:
I don't really, if you had a string longer that 2^64, this code will not work correctly. But, then I guess `encString` needs to be able to fail... Perhaps both encode and decode should be in a monad?